### PR TITLE
[SPARK-11750][SQL] revert SPARK-11727 and code clean up

### PR DESCRIPTION
**Pull request by cloud-fan**
After some experiment, I found it's not convenient to have separate encoder builders: `FlatEncoder` and `ProductEncoder`. For example, when create encoders for `ScalaUDF`, we have no idea if the type `T` is flat or not. So I revert the splitting change in https://github.com/apache/spark/pull/9693, while still keeping the bug fixes and tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 1]
```diff
@@ -718,39 +630,4 @@ trait ScalaReflection {
         throw new UnsupportedOperationException(s"Schema for type $other is not supported")
     }
   }
-
-  def typeOfObject: PartialFunction[Any, DataType] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: This method is only used in tests, so I think we should remove it and its related tests.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala [hunk 2]
```diff
@@ -288,14 +287,13 @@ case class WrapOption(optionType: DataType, child: Expression)
...
   override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
-    val javaType = ctx.javaType(optionType)
     val inputObject = child.gen(ctx)
 
     s"""
       ${inputObject.code}
 
       boolean ${ev.isNull} = false;
-      scala.Option<$javaType> ${ev.value} =
+      scala.Option ${ev.value} =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala]: Janino doesn't support generic type, so we don't need to pass in the `optionType`.

**Comment by cloud-fan** [comment]
cc @marmbrus

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 3]
```diff
@@ -18,34 +18,22 @@
...
-
-/**
- * Support for generating catalyst schemas for scala objects.
- */
-trait ScalaReflection {
-  /** The universe we work in (runtime or macro) */
-  val universe: scala.reflect.api.Universe
-
-  /** The mirror used to access types in the universe */
-  def mirror: universe.Mirror
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: This level of indirection is on purpose.  It lets us use the same reflection code both at runtime and in macros.  Even if we aren't using that right now, we shouldn't undo the work that was done to make this possible.

**Comment by cloud-fan** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 4]
```diff
@@ -736,39 +645,4 @@ trait ScalaReflection {
     assert(methods.length == 1)
     methods.head.getParameterTypes
   }
-
-  def typeOfObject: PartialFunction[Any, DataType] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: This method is only used in tests, so I think we should remove it and its related tests.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 5]
```diff
@@ -44,20 +44,21 @@ import org.apache.spark.sql.types.{NullType, StructField, ObjectType, StructType
  *    to the name `value`.
  */
 object ExpressionEncoder {
-  def apply[T : TypeTag](flat: Boolean = false): ExpressionEncoder[T] = {
+  def apply[T : TypeTag](): ExpressionEncoder[T] = {
     // We convert the not-serializable TypeTag into StructType and ClassTag.
     val mirror = typeTag[T].mirror
     val cls = mirror.runtimeClass(typeTag[T].tpe)
+    val flat = !classOf[Product].isAssignableFrom(cls)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: This isn't going to work for java.  We should just leave this as it was.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala (hunk 4 above)
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: Probably fine in this case, but this isn't related to the undoing of the refactoring and mixing concerns makes the PR much harder to review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 5 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: But we need a way to detect flatness automatically, sometimes we only have the type `T` and nothing else, like when we wanna use encoders for `ScalaUDF`. How about `flat = !ScalaReflection.schemaFor[T].isInstanceOf[StructType])`?

**Pull request closed without merging**